### ISSUE TITLE
Get All Directory needs a manual recursion fallback if the server doesn't support recursion

### ISSUE THREAD

**Issue by datar12345** [robinrodricks/FluentFTP#248]
I having issues.
How to get all directory, subdirectory and file from path root.

ex:
using: ftpClient.GetListing("home/")
result: "home/dir1"
           "home/dir2"
           "home/dir1/csv"
           "home/dir2/csv"
           "home/dir1/csv/test.txt"
           "home/dir2/csv/test.txt"

**Comment by robinrodricks** [comment]
You can add flags to getlisting, check the other overloads.

**Comment by datar12345** [comment]
I use ASP .NET CORE.
I tried to use FluentFTP as after:

var ftpClient = new FtpClient(Configuration["FTPServices:Host"]
                            , Configuration["FTPServices:User"]
                            , Configuration["FTPServices:Password"]) {
                    DataConnectionType = FtpDataConnectionType.AutoActive,
                    DataConnectionEncryption = false,
                    EncryptionMode = FtpEncryptionMode.None,
                    Encoding = System.Text.Encoding.UTF8,
                });

var list = ftpClient.GetListing("home/", FtpListOption.AllFiles | FtpListOption.ForceNameList)
// or -> var list = ftpClient.GetListing("home/", FtpListOption.AllFiles)
// or -> var list = ftpClient.GetListing("home/", FtpListOption.ForceList)

-> But it only returns 2 paths("home/dir1", "home/dir2"), not 6 paths.

I tried to use FtpListOption.UseLS, but an error occurred: command unknown.

May, server not support MLSD, LS command.

Is there any other way.

Please guide help me.

**Comment by robinrodricks** [comment]
Can you remove all your options and only add ftplistoption.recursive??

**Comment by datar12345** [comment]
I tried, but not successful.

I know that the dir -r command is used to get lists all files in current directory and subdirectories.
And i tried use cmd win 7 to connect ftp server and use dir -r command, but it only returns 2 paths("home/dir1", "home/dir2").

Is this the cause I not success with ftplistoption.recursive?

I think you know lot about FTP, so can you explain help me.

**Comment by beantaxi** [comment]
I am having the same issue ... 

ftpClient.GetListing("/", FtpListOption.Recursive);

Only gets me the top level. This is not specific to '/' - I tried other folders and got the same result.

I am very new to FluentFtp, so no clue if this is a regression on not. Any more info I can provide to help?

**Comment by diegoaraujolima** [comment]
I am having the same issue....

**Comment by robinrodricks** [comment]
Guys which servers are you using?
@beantaxi @diegoaraujolima @datar12345

**Comment by diegoaraujolima** [comment]
In my case, IIS 8.

 

De: Robin Rodricks [mailto:notifications@github.com] 
Enviada em: segunda-feira, 4 de dezembro de 2017 05:47
Para: robinrodricks/FluentFTP <FluentFTP@noreply.github.com>
Cc: Diego <diegolima.analista@gmail.com>; Mention <mention@noreply.github.com>
Assunto: Re: [robinrodricks/FluentFTP] Get All Directory (#248)

 

Guys which servers are you using?
@beantaxi <https://github.com/beantaxi>  @diegoaraujolima <https://github.com/diegoaraujolima>  @datar12345 <https://github.com/datar12345>

**Comment by nikitozeg** [comment]
same problem

**Comment by robinrodricks** [comment]
Try this:

conn.GetListing("/",FtpListOption.AllFiles);

**Comment by beantaxi** [comment]
No difference unfortunately.

> Try this:
> 
> conn.GetListing("/",FtpListOption.AllFiles);

**Comment by robinrodricks** [comment]
I think we'll have to add a recursive method to GetListing, which works even if the server didn't support it.

**Comment by FilRav** [comment]
Same issue with linux server. I added this in FtpClient_HighLevel.cs, but it creates too many handles and it crushes sometime... so I'll write not recursively using a stack:
```
    public FtpListItem[] GetListingRecursive(string remotePath)
    {
            if (!remotePath.EndsWith("/")) remotePath += "/";
            FtpListItem[] items = GetListing(remotePath);
            List<FtpListItem> mydir = items.ToList();

            foreach (FtpListItem item in items)
            {
                if (item.Type == FtpFileSystemObjectType.Directory)
                {
                    FtpListItem[] otherItems = GetListingRecursive(item.FullName);  // <- recursive call
                    mydir.AddRange(otherItems.ToList());
                    otherItems = null; // cerco di liberare handle
                }
            }

            return mydir.ToArray();
     }

```
You can add the second parameter "FtpListOption options" (adding also in recursive call obviously).

**Comment by nikitozeg** [comment]
i did the same recursive method 
works perfectly

```
        /// <summary>
        /// Load files from FTP server
        /// </summary>
        public void LoadFtpFiles(string FolderToLoad)
        {
            client.Connect(); 

            foreach (FtpListItem item in client.GetListing(FolderToLoad))
            {
                if (item.Type == FtpFileSystemObjectType.Directory)
                {
                    client.SetWorkingDirectory(item.FullName);
                    LoadFtpFiles(item.FullName);
                }
                else
                    client.DownloadFile(LoadTo + item.FullName, item.FullName);
            }

            client.Disconnect();
        }
```

**Comment by FilRav** [comment]
That's my not-recursive version, it seems to me that handles grow more slowly, maybe I can avoid the crash (for many folders with subfolders and 10K+ files).
```
        public FtpListItem[] GetListingRecursive(string remotePath)
            Stack<string> stack = new Stack<string>();
            stack.Push(remotePath);
            List<FtpListItem> mydir = new List<FtpListItem>();
            while (stack.Count > 0)
            {
                remotePath = stack.Pop();
                if (!remotePath.EndsWith("/")) remotePath += "/";
                FtpListItem[] items = GetListing(remotePath);
                mydir.AddRange(items.ToList());

                foreach (FtpListItem item in items)
                {
                    if (item.Type == FtpFileSystemObjectType.Directory)
                    {
                        stack.Push(item.FullName);
                    }
                }
                items = null;
            }
            return mydir.ToArray();
        }
```

**Comment by rducom** [comment]
Same, I have had to implement manual recursion.
`RecursiveList == true`, but `FtpListOption.Recursive` does not work at all, and  `FtpListOption.AllFiles | FtpListOption.Size | FtpListOption.Recursive` doesn't work either...

I'm facing another issue while doing manual recursion as reported in https://github.com/robinrodricks/FluentFTP/issues/330

**Comment by davidbuckleyni** [comment]
Has this been fixed as I am facing the same issue on the lattest release

**Comment by robinrodricks** [comment]
No its not. We'll work on it.

**Comment by robinrodricks** [comment]
I have added support to detect the FTP server software, and to whitelist specific server types that support recursive file listing. Now all that is remaining is to implement our own manual recursion for servers that do not.

**Comment by robinrodricks** [comment]
I have added support for manual recursion. @all is there anyone to test it?

**Comment by robinrodricks** [comment]
Released in [FluentFTP 20.0.0](https://www.nuget.org/packages/FluentFTP/20.0.0).
